### PR TITLE
Created TF module to perform AMI snapshots using Lamda Functions

### PR DESCRIPTION
**Pull request by s2504s**
## What
* Created TF module `module tf_lambda_ami_backup`

## Why
* This module contains a terraform configuration that creates two lambda functions
that will take AMI automatic at regular intervals.

## References
* Based on https://gist.github.com/bkozora/724e01903a9ad481d21e

### CODE REVIEW

**Review by osterman — CHANGES_REQUESTED** [review]
Can we keep the original filenames? not fond of camelcase filenames for shell scripts.  Also, when renaming files, make sure you use `git mv` so that the diff is cleaner.

File: backup_function.tf [hunk 1]
```diff
@@ -33,95 +33,87 @@ resource "aws_iam_role_policy" "ebs_backup_policy" {
...
 }
 EOF
 }
 
-data "archive_file" "schedule_ebs_snapshot_backups_zip" {
+data "archive_file" "ami_backups_zip" {
   type        = "zip"
-  source_file = "${path.module}/schedule-ebs-snapshot-backups.py"
-  output_path = "${path.module}/schedule-ebs-snapshot-backups.zip"
+  source_file = "lambdaAMIBackups.py"
```
Comment (osterman) [inline backup_function.tf]: use prothon notation (`_`) instead of snakeCase.

File: README.md [hunk 2]
```diff
@@ -15,15 +15,18 @@ Include this repository as a module in your existing terraform code:
 module "lambda_ami_backup" {
   source = "github.com/cloudposse/tf_lambda_ami_backup"
   # Setting these variables is optional
-  # ami_backups_schedule = "cron(00 19 * * ? *)"
-  # ami_cleanups_schedule = "cron(05 19 * * ? *)"
+  ami_backups_schedule = "cron(00 19 * * ? *)"
```
Comment (comeanother) [inline README.md]: Let's use variables for cron schedule.

File: README.md [hunk 3]
```diff
@@ -10,18 +10,23 @@ the code at
 
 Include this repository as a module in your existing terraform code:
 
+
 ```
-module "lambda_ebs_snapshot" {
-  source = "github.com/chef/lambda_ebs_snapshot"
+module "lambda_ami_backup" {
+  source = "github.com/cloudposse/tf_lambda_ami_backup"
```
Comment (osterman) [inline README.md]: Use precise notation.

File: backup_function.tf [hunk 4]
```diff
@@ -1,5 +1,5 @@
-resource "aws_iam_role" "ebs_backup_role" {
-  name = "ebs_backup_role"
+resource "aws_iam_role" "ami_backup" {
+  name = "${module.tf_label.id}"
 
   assume_role_policy = <<EOF
 {
```
Comment (osterman) [inline backup_function.tf]: Do not use HEREDOC

File: main.tf [hunk 5]
```diff
@@ -0,0 +1,6 @@
+module "tf_label" {
+  source    = "github.com/cloudposse/tf_label"
```
Comment (osterman) [inline main.tf]: Use precise format. Pin version to 0.1.0.

**Review by osterman — CHANGES_REQUESTED** [review]
See comments

File: Makefile [hunk 6]
```diff
@@ -1,4 +1,4 @@
-all: ebs-snapshot-janitor.zip schedule-ebs-snapshot-backups.zip
+all: lambda_ami_backups.zip lambda_ami_cleanups.zip
```
Comment (osterman) [inline Makefile:1]: We should add `*.zip` to the `.gitignore`

File: README.md [hunk 7]
```diff
@@ -10,18 +10,23 @@ the code at
...
   # Setting these variables is optional
-  # ebs_snapshot_backups_schedule = "cron(00 19 * * ? *)"
-  # ebs_snapshot_janitor_schedule = "cron(05 19 * * ? *)"
+  ami_backups_schedule = "cron(00 19 * * ? *)"
```
Comment (osterman) [inline README.md]: Normalize spacing

File: backup_function.tf [hunk 8]
```diff
@@ -18,9 +18,9 @@ resource "aws_iam_role" "ebs_backup_role" {
...
 }
 
-resource "aws_iam_role_policy" "ebs_backup_policy" {
-  name = "ebs_backup_policy"
-  role = "${aws_iam_role.ebs_backup_role.id}"
+resource "aws_iam_role_policy" "ami_backup" {
+  name = "${module.tf_label.id}"
+  role = "${aws_iam_role.ami_backup.id}"
 
   policy = <<EOF
```
Comment (osterman) [inline backup_function.tf]: Do not use HEREDOC

File: variables.tf [hunk 9]
```diff
@@ -1,9 +1,33 @@
 # See https://docs.aws.amazon.com/lambda/latest/dg/tutorial-scheduled-events-schedule-expressions.html
 # for how to write schedule expressions
-variable "ebs_snapshot_backups_schedule" {
+variable "ami_backups_schedule" {
```
Comment (osterman) [inline variables.tf]: Drop `ami_` as it's implied by the name of the module.

File: variables.tf [hunk 10]
```diff
@@ -1,9 +1,33 @@
...
   default = "cron(00 19 * * ? *)"
 }
 
-variable "ebs_snapshot_janitor_schedule" {
+variable "ami_cleanups_schedule" {
```
Comment (osterman) [inline variables.tf]: Drop `ami_` as it's implied by the name of the module.

File: main.tf [hunk 11]
```diff
@@ -1,5 +1,5 @@
 module "tf_label" {
-  source    = "github.com/cloudposse/tf_label"
+  source    = "../tf_label"
```
Comment (const-bon) [inline main.tf]: Replace `../tf_label` by github source

File: main.tf [hunk 12]
```diff
@@ -0,0 +1,6 @@
+module "tf_label" {
+  source    = "git::https://github.com/cloudposse/tf_label.git?ref=0.1.0"
```
Comment (const-bon) [inline main.tf]: Can we use `git::https://github.com/cloudposse/tf_label?ref=0.1.0`? 
Without `.git`

Comment (osterman) [reply main.tf]: i prefer we just standardize around this format:
```
git::https://github.com/cloudposse/tf_label.git?ref=tags/0.1.0
```

Note: be explicit about `tags/` and use the `.git` suffix.

File: variables.tf [hunk 13]
```diff
@@ -1,9 +1,33 @@
 # See https://docs.aws.amazon.com/lambda/latest/dg/tutorial-scheduled-events-schedule-expressions.html
 # for how to write schedule expressions
-variable "ebs_snapshot_backups_schedule" {
+variable "backups_schedule" {
```
Comment (osterman) [inline variables.tf]: lets call it `backup_schedule`

File: variables.tf [hunk 14]
```diff
@@ -1,9 +1,33 @@
...
   default = "cron(00 19 * * ? *)"
 }
 
-variable "ebs_snapshot_janitor_schedule" {
+variable "cleanups_schedule" {
```
Comment (osterman) [inline variables.tf]: Lets call it `cleanup_schedule`

File: Makefile (hunk 6 above)
Comment (const-bon) [reply Makefile:1]: Added

File: README.md [hunk 15]
```diff
@@ -1,7 +1,7 @@
-# Terraform config for automatic EBS snapshots
+# Terraform config for automatic AMI create
```
Comment (const-bon) [inline README.md:1]: Replace to `# Terraform module for automatic scheduled AMI creation`

File: README.md [hunk 16]
```diff
@@ -1,7 +1,7 @@
...
 
 This repo contains a terraform configuration that creates two lambda functions
-that will take automatic EBS snapshots at regular intervals. It is based on
+that will take AMI automatic at regular intervals. It is based on
```
Comment (const-bon) [inline README.md:4]: Replace to `that will create AMI automatically at regular intervals`

File: backup_function.tf [hunk 17]
```diff
@@ -1,127 +1,138 @@
...
+    ]
+
+    resources = [
+      "arn:aws:logs:*:*:*",
+    ]
+  }
+
+  statement {
+    actions = [
+      "ec2:*",
```
Comment (osterman) [inline backup_function.tf:32]: Why are we giving carte blanche access? Please use explicit permissions and not `*`

File: backup_function.tf [hunk 18]
```diff
@@ -1,127 +1,138 @@
...
+  source_file = "${path.module}/lambda_ami_backups.py"
+  output_path = "${path.module}/lambda_ami_backups.zip"
 }
 
-resource "aws_lambda_function" "schedule_ebs_snapshot_backups" {
-  filename         = "${path.module}/schedule-ebs-snapshot-backups.zip"
-  function_name    = "schedule_ebs_snapshot_backups"
+resource "aws_lambda_function" "lambda_ami_backups" {
+  filename         = "${path.module}/lambda_ami_backups.zip"
+  function_name    = "${module.tf_label.id}_backups"
```
Comment (osterman) [inline backup_function.tf:60]: label module by default uses `-` and not `_`. Let's use `-` here.

File: backup_function.tf [hunk 19]
```diff
@@ -1,127 +1,138 @@
...
 }
 
-resource "aws_lambda_function" "ebs_snapshot_janitor" {
-  filename         = "${path.module}/ebs-snapshot-janitor.zip"
-  function_name    = "ebs_snapshot_janitor"
-  description      = "Cleans up old EBS backups"
-  role             = "${aws_iam_role.ebs_backup_role.arn}"
+resource "aws_lambda_function" "lambda_ami_cleanups" {
+  filename         = "${path.module}/lambda_ami_cleanups.zip"
+  function_name    = "${module.tf_label.id}_cleanups"
```
Comment (osterman) [inline backup_function.tf:84]: lets use `-` here.

File: backup_function.tf [hunk 20]
```diff
@@ -1,127 +1,138 @@
...
-  function_name    = "schedule_ebs_snapshot_backups"
+resource "aws_lambda_function" "lambda_ami_backups" {
+  filename         = "${path.module}/lambda_ami_backups.zip"
+  function_name    = "${module.tf_label.id}_backups"
   description      = "Automatically backs up instances tagged with backup: true"
-  role             = "${aws_iam_role.ebs_backup_role.arn}"
+  role             = "${aws_iam_role.ami_backup.arn}"
   timeout          = 60
-  handler          = "schedule-ebs-snapshot-backups.lambda_handler"
+  handler          = "lambda_ami_backups.lambda_handler"
```
Comment (osterman) [inline backup_function.tf:64]: lets use the singular form (`backup`) consistently.

**Comment by const-bon** [comment]
Let's close this PR in favor of https://github.com/cloudposse/tf_lambda_ami_backup/pull/2.

**Comment by osterman** [comment]
Closed in favor of https://github.com/cloudposse/tf_lambda_ami_backup/pull/2

**Pull request closed without merging**